plugins/event/sqs/sqs.go

```diff
@@ -0,0 +1,256 @@
+	}
+	s.connections[request.SubscriptionID] = conn
+	return nil
+}
+
+func (s *sqsBackend) killConnection(subscriptionID string) {
+	s.clientLock.Lock()
+	defer s.clientLock.Unlock()
+	s.killConnectionWithLock(subscriptionID)
+}
+
+func (s *sqsBackend) killConnectionWithLock(subscriptionID string) {
+	conn, ok := s.connections[subscriptionID]
+	if !ok {
+		return
+	}
+	conn.ctxCancelFunc()
+	delete(s.connections, subscriptionID)
+}
+
+func (s *sqsBackend) getConn(subscriptionID string) (*sqsConnection, error) {
+	s.clientLock.RLock()
+	defer s.clientLock.RUnlock()
+	conn, ok := s.connections[subscriptionID]
+	if !ok {
+		return nil, fmt.Errorf("invalid subscription_id")
+	}
+	return conn, nil
+}
+
+func (s *sqsBackend) Send(ctx context.Context, request *event.Request) error {
+	if request.Subscribe != nil {
+		return s.subscribe(ctx, request.Subscribe)
+	} else if request.Unsubscribe != nil {
+		return s.Unsubscribe(ctx, request.Unsubscribe.SubscriptionID)
+	} else {
+		return s.sendEvent(ctx, request.Event)
+	}
+}
+
+func (s *sqsBackend) sendEvent(_ context.Context, send *event.SendEventRequest) error {
+	return s.sendSubscriptionEventInternal(send.SubscriptionID, send.EventJSON, false)
+}
+
+func (s *sqsBackend) refreshClient(subscriptionID string) error {
+	conn, err := s.getConn(subscriptionID)
+	if err != nil {
+		return err
+	}
+	client, err := newClient(conn.config)
+	if err != nil {
+		return err
+	}
+	s.clientLock.Lock()
+	defer s.clientLock.Unlock()
+	conn.client = client
+	// probably not necessary, but just in case
+	s.connections[subscriptionID] = conn
+	return nil
+}
+
+func (s *sqsBackend) sendSubscriptionEventInternal(subscriptionID string, eventJson string, isRetry bool) error {
+	conn, err := s.getConn(subscriptionID)
+	if err != nil {
+		return err
+	}
+	backoff := conn.config.NewRetryBackoff()
+	err = backoff.Retry(func() error {
+		_, err = conn.client.SendMessage(&sqs.SendMessageInput{
+			MessageBody: &eventJson,
+			QueueUrl:    &conn.queueURL,
+		})
+		return err
+	})
+	if err != nil && !isRetry {
+		// refresh client and try again, once
+		err2 := s.refreshClient(subscriptionID)
+		if err2 != nil {
+			return errors.Join(err, err2)
+		}
+		return s.sendSubscriptionEventInternal(subscriptionID, eventJson, true)
+	} else if err != nil && isRetry {
+		s.killConnection(subscriptionID)
+		return err
+	}
+	return nil
+}
+
+func (s *sqsBackend) Unsubscribe(_ context.Context, subscriptionID string) error {
+	s.killConnection(subscriptionID)
+	return nil
+}
+
+func (s *sqsBackend) PluginName() string {
+	return pluginName
+}
+
+func (s *sqsBackend) PluginVersion() logical.PluginVersion {
+	return logical.PluginVersion{
+		Version: version.GetVersion().Version,
+	}
+}
+
+func (s *sqsBackend) Close(_ context.Context) error {
+	s.clientLock.Lock()
+	defer s.clientLock.Unlock()
+	var subscriptions []string
+	for k := range s.connections {
+		subscriptions = append(subscriptions, k)
+	}
+	for _, subscription := range subscriptions {
```

Comment:
The kill connection with lock modifies the map, so the first loop would not be safe to iterate through. (I *think* it might be safe, but I am always 👀 when I see modifying the map you're iterating through.)

Reply:
Oh yeah of course 👍 thanks :) and definitely agreed on not modifying a map while you iterate through it